lib/sparrow/pools_warden.ex

```diff
@@ -94,33 +94,43 @@ defmodule Sparrow.PoolsWarden do
       end)
   end
 
-  @spec handle_info(any, :ok) :: {:noreply, :ok}
-  def handle_info(unknown, _state) do
+  @impl GenServer
+  def handle_info({:DOWN, _ref, :process, pid, reason}, state) do
+    {pid, pool_type, pool_name, tags} = List.keyfind(state, pid, 0)
+    :ets.delete_object(@tab_name, {pool_type, {pool_name, tags}})
+
+    _ =
+      Logger.info(fn ->
+        "worker=pools_warden, action=unregistering pool, pool_name=#{inspect(pool_name)}, pid=#{inspect(pid)},
+        reason=#{inspect(reason)}"
+      end)
+
+
+    {:noreply, List.delete(state, {pid, pool_type, pool_name, tags})}
+  end
+
+
+  def handle_info(unknown, state) do
     _ =
       Logger.warn(fn ->
         "worker=pools_warden, Unknown info #{inspect(unknown)}"
       end)
 
-    {:noreply, :ok}
+    {:noreply, state}
   end
 
-  @spec handle_call(
-          {:add_pool, pool_type, atom, [any]},
-          GenServer.from(),
-          any
-        ) :: {:reply, atom, :ok}
-
-  def handle_call({:add_pool, pool_type, pool_name, tags}, _, _state) do
+  @impl GenServer
+  def handle_call({:add_pool, pid, pool_type, pool_name, tags}, _, state) do
     :ets.insert(@tab_name, {pool_type, {pool_name, tags}})
+    Process.monitor(pid)
 
     _ =
       Logger.info(fn ->
         "worker=pools_warden, action=adding_pool, pool_type=#{
           inspect(pool_type)
         }, pool_name=#{inspect(pool_name)}, pool_tags=#{inspect(tags)}"
       end)
-
-    {:reply, pool_name, :ok}
+    {:reply, pool_name, [{pid, pool_type, pool_name, tags} | state]}
```

Comment:
You use `List.keyfind/3` to search through state, but your list allows for duplicates of `pid` key.